pandas/core/indexes/datetimelike.py

```diff
@@ -854,6 +854,116 @@ def _concat_same_dtype(self, to_concat, name):
+
+    def _intersect_ascending(self, other, **empty_params):
+        # to make our life easier, "sort" the two ranges
+        if self[0] <= other[0]:
+            left, right = self, other
+        else:
+            left, right = other, self
+
+        end = min(left[-1], right[-1])
+        start = right[0]
+
+        if end < start:
+            return type(self)(data=[], **empty_params)
+        return left._slice_by_value(start, end)
+
+    def _intersect_descending(self, other, **empty_params):
+        # this is essentially a flip of _intersect_ascending
+        if self[0] >= other[0]:
+            left, right = self, other
+        else:
+            left, right = other, self
+
+        start = min(left[0], right[0])
+        end = right[-1]
+
+        if end > start:
+            return type(self)(data=[], **empty_params)
+        return left._slice_by_value(start, end)
+
+    def _offsets_equal(self, other):
+        offsets_equal = True
+        self_offset = getattr(self, 'offset', None)
+        other_offset = getattr(other, 'offset', None)
+        if self_offset is None or other_offset is None or \
+                self_offset != other_offset or other_offset.isAnchored():
+            offsets_equal = False
+        return offsets_equal
+
+    def intersection(self, other):
+        """
+        Specialized intersection for DateTimeIndexOpsMixin objects.
+        May be much faster than Index.intersection.
+
+        Parameters
+        ----------
+        other : Index or array-like
+
+        Returns
+        -------
+        Index
+            A shallow copied intersection between the two things passed in
+        """
+        self._assert_can_do_setop(other)
+
+        if self.equals(other):
+            return self._get_consensus_name(other)
+
+        if not isinstance(other, DatetimeIndexOpsMixin):
+            try:
+                other = DatetimeIndexOpsMixin(other)
+            except (TypeError, ValueError):
+                pass
+            result = Index.intersection(self, other)
+            return result
+        elif (self._offsets_equal(other) or
+              (not self.is_strictly_monotonic or
+                not other.is_strictly_monotonic)):
+            result = Index.intersection(self, other)
+            tz = getattr(self, 'tz', None)
+            result = self._shallow_copy(result._values, name=result.name,
+                                        tz=tz, freq=None)
+            if result.freq is None:
+                result.offset = frequencies.to_offset(result.inferred_freq)
+            return result
+
+        lenghts = len(self), len(other)
+        if lenghts[0] == 0:
+            return self
+        if lenghts[1] == 0:
+            return other
+
+        # handle intersecting things like this
+        # idx1 = pd.to_timedelta((1, 2, 3, 4, 5, 6, 7, 8), unit='s')
+        # idx2 = pd.to_timedelta((2, 3, 4, 8), unit='s')
+        if lenghts[0] != lenghts[1] and (
```

Comment:
why do you need this special case?

Reply:
The way that fast indexing works will produce a wrong result when running an intersection with those two examples. The result will be either (2, 3, 4, 8) or (1, 2, 3, 4, 5, 6, 6, 8) depending on which index called intersection.

Interestingly enough, when this case is gone, the reason I added the assert kind in ['ix', 'loc', 'getitem', None] shows up as n error in one of the tests. With this case it goes away, probably because that case was being called

- pandas/tests/indexes/period/test_setops.py:133 TestPeriodIndex.test_intersection: AssertionError
- pandas/tests/indexes/period/test_setops.py:155 TestPeriodIndex.test_intersection_cases: AssertionError

I added None as None is in both the base _maybe_cast_slice_bound and the timedeltas _maybe_cast_slice_bound.